tests/core/io/test_tcp_server.h

```diff
@@ -0,0 +1,216 @@
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef TEST_TCP_SERVER_H
+#define TEST_TCP_SERVER_H
+
+#include "core/io/stream_peer_tcp.h"
+#include "core/io/tcp_server.h"
+#include "tests/test_macros.h"
+
+namespace TestTCPServer {
+
+const int port = 12345;
+const IPAddress localhost("127.0.0.1");
+
+Ref<TCPServer> create_server(const IPAddress &p_address, int p_port) {
+	Ref<TCPServer> server;
+	server.instantiate();
+
+	Error err = server->listen(port, localhost);
+	REQUIRE_EQ(Error::OK, err);
+	REQUIRE(server->is_listening());
+	CHECK_FALSE(server->is_connection_available());
+
+	return server;
+}
+
+Ref<StreamPeerTCP> create_client(const IPAddress &p_address, int p_port) {
+	Ref<StreamPeerTCP> client;
+	client.instantiate();
+
+	Error err = client->connect_to_host(localhost, port);
+	REQUIRE_EQ(Error::OK, err);
+	CHECK_EQ(localhost, client->get_connected_host());
+	CHECK_EQ(port, client->get_connected_port());
+	CHECK_EQ(StreamPeerTCP::STATUS_CONNECTING, client->get_status());
+
+	return client;
+}
+
+Ref<StreamPeerTCP> accept_connection(Ref<TCPServer> &server) {
+	REQUIRE(server->is_connection_available());
+	Ref<StreamPeerTCP> client_from_server = server->take_connection();
+	REQUIRE_MESSAGE(client_from_server.is_valid(), "A Stream Peer TCP from the TCP Server should not be a null pointer.");
+	CHECK_EQ(localhost, client_from_server->get_connected_host());
+	CHECK_EQ(StreamPeerTCP::STATUS_CONNECTED, client_from_server->get_status());
+
+	return client_from_server;
+}
+
+TEST_CASE("[TCPServer] Instantiation") {
+	Ref<TCPServer> server;
+	server.instantiate();
+
+	REQUIRE_MESSAGE(server.is_valid(), "A TCP Server created should not be a null pointer.");
+	CHECK_EQ(server->is_listening(), false);
+}
+
+TEST_CASE("[TCPServer] Accept a connection and receive/send data") {
+	Ref<TCPServer> server = create_server(localhost, port);
+	Ref<StreamPeerTCP> client = create_client(localhost, port);
+
+	// Required to get the connection properly established.
+	OS::get_singleton()->delay_usec(500000);
```

Comment:
It's always recommended to `poll` while waiting

Reply:
Could you please elaborate a little bit more?
A couple of lines below I'm calling `client->poll()`, so maybe I'm not fully getting how should I use `poll`.

Reply:
@pafuent Great work and thanks for working on this! Not Fabio, and he may correct me, but:

Polling is what allows the connection (or underlying socket) to execute and do actual work. They don't run by themselves in parallel in threads or anything, they literally only do work when polled.

The server polls its socket when calling `server->is_connection_available()` (see [here](https://github.com/godotengine/godot/blob/e3213aaef5e0e72b8272e65d989d3d8222be17ca/core/io/tcp_server.cpp#L97); I think we should actually point this one out more directly in TCPServer [docs](https://docs.godotengine.org/en/stable/classes/class_tcpserver.html#class-tcpserver-method-is-connection-available), looking at it), the client as you point out with `client->poll()`.

This means that it is good practice to poll as regularly as possible, because the longer the time between polls:
- the higher the chance networking buffers may fill up/overrun and you lose packets/get errors
- the higher the chance you process messages too late (if you were to poll, say, every five seconds, you may end up processing information from 4 seconds ago that is by now outdated and no longer relevant, you don't really need the player's movement update from 4 seconds ago)
- the longer the connection appears to "hang" for the other side, as they won't get a packet response until you poll
- as per the above, this also introduces problems in transit, especially for TCP (which is not "fire and forget" like UDP), because both your node, the remote node and any node in between could decide your connection is dead because there has been no traffic, so why not kill the connection to free up resources? I.e., timeouts/connection terminations may happen, usually only with very high poll times, of course :)

Of course, all this is a bit more theoretical and "best practice" than 100% applicable to a test case like this. *However*, looking at it I am a bit surprised it works by just literally polling server and client once each. Instead of waiting arbitrarily for half a second, you could replace that with a loop that sleeps, say, 1000usecs, and then polls both client and server each time until you get your desired result (`StreamPeerTCP::STATUS_CONNECTED`), or you run longer then excpected and time out/fail the test.

Reply:
Thanks for the amazing explanation!!!
In my case I needed to add that sleep because when I called `poll` or other method that does that internally I was getting a wrong return code or value. IIRC that was happening on Windows, but not on Mac or Linux.
Besides of that, if you think that it's worth it for the tests,  I can try to change them to poll on a for loop with a smaller sleep time.

Reply:
> In my case I needed to add that sleep because when I called `poll` or other method that does that internally I was getting a wrong return code or value.

Indeed, the problem here is that you keep blocking for a long time and only poll once, while you should only block for brief periods of time and regularly poll, or you risk of having the same error on other OS configurations.